### PR TITLE
Bpb359: Intermediate do not merge

### PR DESCRIPTION
**Pull request by scribe**
This isn't perfect but it makes it possible to exit a transfer job early.
This also fixes an issue with paths more than the OS ulimit deep.

### CODE REVIEW

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/GetJob.kt [hunk 1]
```diff
@@ -80,7 +81,7 @@ class GetJob(private val getJobData: JobData) : JobService(), PrepStage<JobData>
         getJobData.removeJob()
     }
 
-    override fun finishedCompletable(): Completable {
+    override fun finishedCompletable(cancelled: KFunction0<Boolean>): Completable {
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/GetJob.kt]: I don't see this function being called here.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobFacade.kt [hunk 2]
```diff
@@ -17,19 +17,17 @@ package com.spectralogic.dsbrowser.gui.services.jobService
...
-import java.util.concurrent.atomic.AtomicLong
+import kotlin.reflect.KFunction0
 
 interface JobFacade {
     public fun titleObservable(): Observable<String>
     public fun jobSizeObservable(): Observable<Number>
     public fun messageObservable(): Observable<String>
     public fun visabilityObservable(): Observable<Boolean>
-    public fun finishedCompletable(): Completable
+    public fun finishedCompletable(cancelled: KFunction0<Boolean>): Completable
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobFacade.kt]: Why are you using a `KFunction0<Boolean>` instead of a `Consumer<Boolean>`

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/GetJobData.kt [hunk 3]
```diff
@@ -39,11 +39,14 @@ import java.nio.file.Path
...
 import java.time.Instant
 import java.util.*
+import kotlin.reflect.KFunction
+import kotlin.reflect.KFunction0
 
 data class GetJobData(private val list: List<Pair<String, String>>,
                       private val localPath: Path,
                       override val bucket: String,
                       private val jobTaskElement: JobTaskElement) : JobData {
+    override var cancelled: KFunction0<Boolean>? = null
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/GetJobData.kt]: Why is this here? and set to null?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/JobData.kt [hunk 4]
```diff
@@ -48,4 +50,5 @@ interface JobData {
     var prefixMap: MutableMap<String, Path>
     public var lastFile: String
     public var jobId: UUID?
+    public var cancelled: KFunction0<Boolean>?
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/JobData.kt]: I'm a little confused by this design.  Why is this not a boolean property?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt [hunk 5]
```diff
@@ -96,18 +104,22 @@ data class PutJobData(private val items: List<Pair<String, Path>>,
...
+                                , if (Files.isDirectory(it)) 0L else Files.size(it))
                     }
-                            , if (Files.isDirectory(it)) 0L else Files.size(it))
-                }.collect(GuavaCollectors.immutableList())
-        ds3ObjectList.forEach({ prefixMap.put(it.name, item.second) })
-        return ds3ObjectList
+                    .toList()
+        }
+        paths.forEach { prefixMap.put(it.name, item.second) }
+        return paths.asIterable()
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt]: Isn't a list already an iterable?  Why this call?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobFacade.kt (hunk 2 above)
Comment (scribe) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobFacade.kt]: That's what Intelij typed it as, I'll change it.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt (hunk 5 above)
Comment (scribe) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt]: Good catch.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobFacade.kt (hunk 2 above)
Comment (scribe) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobFacade.kt]: I need the result this returns, Consumer doesn't give me back the value.

Comment (rpmoore) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobFacade.kt]: Then use a Supplier

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/GetJob.kt [hunk 6]
```diff
@@ -28,6 +28,9 @@ import com.spectralogic.dsbrowser.gui.util.toByteRepresentation
 import io.reactivex.Completable
 import org.slf4j.LoggerFactory
 import java.util.*
+import java.util.function.Consumer
+import java.util.function.Supplier
+import kotlin.reflect.KFunction0
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/GetJob.kt]: It looks like a couple of these imports were left in by accident.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/GetJob.kt [hunk 7]
```diff
@@ -80,11 +83,15 @@ class GetJob(private val getJobData: JobData) : JobService(), PrepStage<JobData>
         getJobData.removeJob()
     }
 
-    override fun finishedCompletable(): Completable {
+    override fun finishedCompletable(cancelled: Supplier<Boolean>): Completable {
         return Completable.fromAction {
             val prep = prepare(getJobData)
-            transfer(prep)
-            tearDown()
+            if (!cancelled.get()) {
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/GetJob.kt]: To me this would read better if the cancel logic returned instead.  I believe you can do this with kotlin and the `return@fromAction` syntax.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobFacade.kt [hunk 8]
```diff
@@ -17,19 +17,18 @@ package com.spectralogic.dsbrowser.gui.services.jobService
...
 import io.reactivex.Completable
 import io.reactivex.Observable
-import io.reactivex.Observer
 import javafx.beans.property.LongProperty
 import javafx.beans.property.SimpleBooleanProperty
-import java.time.Instant
 import java.util.*
-import java.util.concurrent.atomic.AtomicLong
+import java.util.function.Supplier
+import kotlin.reflect.KFunction0
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobFacade.kt]: Remove this import.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobService.kt [hunk 9]
```diff
@@ -15,12 +15,11 @@
...
 
 import com.github.thomasnield.rxkotlinfx.toObservable
-import com.spectralogic.ds3client.helpers.Ds3ClientHelpers
 import io.reactivex.Completable
 import io.reactivex.Observable
 import javafx.beans.property.*
-import org.reactivestreams.Publisher
-import java.util.*
+import java.util.function.Supplier
+import kotlin.reflect.KFunction0
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobService.kt]: Remove this import.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobTask.kt [hunk 10]
```diff
@@ -75,19 +75,20 @@ class JobTask(private val wrappedJob: JobFacade) : Ds3JobTask() {
...
                            deepStorageBrowserPresenter: DeepStorageBrowserPresenter): (WorkerStateEvent) -> Unit = {
         jobId.exists {
-            try {
-                Platform.runLater {
+            workers.execute {
+                try {
                     client.cancelJobSpectraS3(CancelJobSpectraS3Request(it))
+                    log.info("{} Job cancelled", type)
+                    loggingService.logMessage("$type Job Cancelled", LogType.INFO)
+                } catch (e: IOException) {
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobTask.kt]: What about a call all exception handler?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobTask.kt [hunk 11]
```diff
@@ -96,10 +97,13 @@ class JobTask(private val wrappedJob: JobFacade) : Ds3JobTask() {
                  deepStorageBrowserPresenter: DeepStorageBrowserPresenter,
                  loggingService: LoggingService,
                  log: Logger,
+                 workers: Workers,
                  type: String): (WorkerStateEvent) -> Unit = { worker: WorkerStateEvent ->
         val throwable: Throwable = worker.source.exception
         jobId.exists {
-            ParseJobInterruptionMap.removeJobID(jobInterruptionStore, it.toString(), client.connectionDetails.endpoint, deepStorageBrowserPresenter, loggingService)
+            workers.execute {
+                ParseJobInterruptionMap.removeJobID(jobInterruptionStore, it.toString(), client.connectionDetails.endpoint, deepStorageBrowserPresenter, loggingService)
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobTask.kt]: Can this call throw an exception?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/PutJob.kt [hunk 12]
```diff
@@ -32,24 +32,32 @@ import io.reactivex.disposables.Disposable
...
     override fun jobUUID(): UUID? = putJobData.jobId
 
-    override fun finishedCompletable(): Completable {
+    override fun finishedCompletable(cancelled: Supplier<Boolean>): Completable {
+        putJobData.cancelled = cancelled
         return Completable.fromAction {
             val resources = prepare(putJobData)
-            transfer(resources)
-            tearDown()
+            if (cancelled.get() == false) {
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/PutJob.kt]: Earlier, you had `!cancelled.get()` instead of this explicit comparison.  Pick on or the other.  And I am curious if the alternative `return@` syntax will work here.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/GetJobData.kt [hunk 13]
```diff
@@ -39,11 +39,15 @@ import java.nio.file.Path
 import java.nio.file.Paths
 import java.time.Instant
 import java.util.*
+import java.util.function.Supplier
+import kotlin.reflect.KFunction
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/GetJobData.kt]: Clean up imports.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/JobData.kt [hunk 14]
```diff
@@ -25,6 +25,9 @@ import javafx.beans.property.SimpleBooleanProperty
 import java.nio.file.Path
 import java.time.Instant
 import java.util.*
+import java.util.function.Supplier
+import kotlin.reflect.KFunction
+import kotlin.reflect.KFunction0
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/JobData.kt]: imports

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/GetJobData.kt [hunk 15]
```diff
@@ -55,6 +59,7 @@ data class GetJobData(private val list: List<Pair<String, String>>,
     }
 
 
+
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/GetJobData.kt]: Why the extra white space?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt [hunk 16]
```diff
@@ -26,17 +27,22 @@ import com.spectralogic.ds3client.models.JobStatus
...
 import com.spectralogic.dsbrowser.util.GuavaCollectors
 import javafx.beans.property.BooleanProperty
+import sun.util.calendar.ZoneInfoFile
 import java.nio.file.Files
 import java.nio.file.Path
 import java.time.Instant
 import java.util.*
+import java.util.function.Supplier
+import java.util.stream.Stream
+import kotlin.reflect.KFunction0
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt]: cleanup imports.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt [hunk 17]
```diff
@@ -96,18 +105,22 @@ data class PutJobData(private val items: List<Pair<String, Path>>,
...
         val localDelim = item.second.fileSystem.separator
         val parent = item.second.parent
-        val ds3ObjectList = Files.walk(item.second)
-                .filter { !(Files.isDirectory(it) && Files.list(it).findAny().isPresent) }
-                .map {
-                    Ds3Object(if (Files.isDirectory(it)) {
-                        targetDir + parent.relativize(it).toString().replace(localDelim, "/") + "/"
-                    } else {
-                        targetDir + parent.relativize(it).toString().replace(localDelim, "/")
+        val paths = Files.walk(item.second).use { stream: Stream<Path> ->
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt]: consider pulling this out into its own method.  Also, why return a list?  Doesn't stream have a forEach method on it?  That would avoid having to aggregate it all up into a list to then turn around and re-iterate over it.

Comment (rpmoore) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/data/PutJobData.kt]: Oh nvm, I answered my own question.  This method side effects...

File: dsb-gui/src/test/java/com/spectralogic/dsbrowser/gui/services/jobService/JobServiceTest.kt [hunk 18]
```diff
@@ -23,6 +23,8 @@ import org.junit.Test
 import org.assertj.core.api.Assertions.*
 import org.mockito.Mockito
 import java.util.*
+import java.util.function.Supplier
+import kotlin.reflect.KFunction0
```
Comment (rpmoore) [inline dsb-gui/src/test/java/com/spectralogic/dsbrowser/gui/services/jobService/JobServiceTest.kt]: Cleanup imports.
